### PR TITLE
feat(ST-146): add update user api endpoint

### PR DESCRIPTION
**Pull request by faysal-alam-sazim**
## Description of Change

- Added Update User DTO
- Added PATCH `users/profile` api endpoint
- Added necessary service methods

## Associated Ticket Link(s)

- https://trello.com/c/qRYEYYtI

## Related Pull Request(s)

- N/A

### CODE REVIEW

File: src/users/users.controller.ts [hunk 1]
```diff
@@ -53,4 +53,17 @@ export class UsersController {
 
     return this.usersSerializer.serialize(userProfile);
   }
+
+  @UseGuards(JwtAuthGuard)
+  @Patch("profile")
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.controller.ts:58]: Same comment as previous PR

File: src/users/users.controller.ts [hunk 2]
```diff
@@ -53,4 +53,17 @@ export class UsersController {
...
+  async updateUser(@CurrentUser() user: ITokenizedUser, @Body() updateUserDto: UpdateUserDto) {
+    const updatedUser = await this.usersService.updateUser(user.id, updateUserDto);
+
+    const accessToken = await this.authService.createAccessToken(updatedUser);
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.controller.ts:62]: Why are we creating an accessToken when updating?

Comment (faysal-alam-sazim) [reply src/users/users.controller.ts:62]: I am using first name in the payload of the jwt token. So if first name is changed, I creating another token sending to the client.

Comment (ishmam-iqbal-sazim) [reply src/users/users.controller.ts:62]: 1. If first name is changed, does the previous token get invalidated or unusable?

File: src/users/users.service.ts [hunk 3]
```diff
@@ -80,4 +80,32 @@ export class UsersService {
...
+  async updateUser(id: number, updateUserDto: UpdateUserDto) {
+    const user = await this.usersRepository.findOneOrFail(id);
+    const { firstName, lastName, gender } = updateUserDto;
+
+    this.usersRepository.assign(user, { firstName, lastName, gender });
+
+    if (user.role === EUserRole.TEACHER) {
+      const { teacherInput } = updateUserDto;
+
+      this.entityManager.assign(user.teacher, teacherInput);
```
Comment (ishmam-iqbal-sazim) [inline src/users/users.service.ts:93]:
```suggestion
      this.usersRepository.assign(user, { firstName, lastName, gender, teacher: teacherInput });
```

Can't we do this? 
Same comment for assigning values to user.student

File: src/users/users.controller.ts (hunk 2 above)
Comment (faysal-alam-sazim) [reply src/users/users.controller.ts:62]: 1. Removed access token as it doesn't need to change. Just sending updated user info as tokenized user to store in the redux store.
2. I've already add get user info api endpoint so no need to call after every change getMe

File: src/users/users.service.ts (hunk 3 above)
Comment (faysal-alam-sazim) [reply src/users/users.service.ts:84]: No bhai. I causes an error which is teacher field contains id but we are giving objects to assign. We can't assign the teacher object with the EM of usersRepository.

**Review by ishmam-iqbal-sazim — APPROVED** [review]

**Review by ragib-hossain-sazim — APPROVED** [review]

**Pull request merged**